File: dart/veritag_sdk/lib/src/attest_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Thrown on any non-2xx response, carrying the status and the API's detail.
class AttestApiError implements Exception {
  AttestApiError(this.status, this.detail);
  final int status;
  final String detail;
  @override
  String toString() => 'AttestCore $status: $detail';
}

/// Thin HTTP client for the AttestCore web API. Used by any app in the
/// ecosystem to provision tags and verify scans. Domain-agnostic.
class AttestClient {
  AttestClient(this.baseUrl, {this.bearerToken, http.Client? client})
      : _http = client ?? http.Client();

  final String baseUrl;
  final String? bearerToken;
  final http.Client _http;

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (bearerToken != null) 'Authorization': 'Bearer $bearerToken',
      };

  /// Decode a response, throwing [AttestApiError] on non-2xx instead of casting
  /// an error body to a success type.
  Map<String, dynamic> _parse(http.Response r) {
    final body = r.body.isEmpty
        ? <String, dynamic>{}
        : jsonDecode(r.body) as Map<String, dynamic>;
    if (r.statusCode ~/ 100 != 2) {
      throw AttestApiError(r.statusCode, (body['detail'] ?? r.reasonPhrase ?? '') as String);
    }
    return body;
  }

  /// Register an issuer. Returns its id AND the bearer token (shown once) needed
  /// to authenticate provisioning.
  Future<({String issuerId, String token})> registerIssuer(String publicKeyHex,
      {Map<String, dynamic> policy = const {}}) async {
    final r = await _http.post(Uri.parse('$baseUrl/v1/issuers'),
        headers: _headers,
        body: jsonEncode({'public_key_hex': publicKeyHex, 'policy': policy}));
    final b = _parse(r);
    return (issuerId: b['issuer_id'] as String, token: b['token'] as String);
  }

  /// Revoke an issuer (an issuer may only revoke itself).
  Future<void> revokeIssuer(String issuerId) async {
    final r = await _http.post(
        Uri.parse('$baseUrl/v1/issuers/${Uri.encodeComponent(issuerId)}/revoke'),
        headers: _headers);
    _parse(r);
  }

  Future<Map<String, dynamic>> provision({
    required String uid,
    required String issuerId,
    required String bindingPayloadHex,
    required String signatureHex,
    int keyVersion = 1,
  }) async {
    final r = await _http.post(Uri.parse('$baseUrl/v1/tags/provision'),
        headers: _headers,
        body: jsonEncode({
          'uid': uid,
          'issuer_id': issuerId,
          'binding_payload_hex': bindingPayloadHex,
          'signature_hex': signatureHex,
          'key_version': keyVersion,
        }));
    return _parse(r);
  }

  /// Verify a scan. Returns the three independent booleans from the core.
  Future<VerifyResult> verify(String uid, int ctr, String cmacHex,
      {int keyVersion = 1}) async {
    final uri = Uri.parse('$baseUrl/v1/verify').replace(queryParameters: {
      'u': uid,
      'c': '$ctr',
      'm': cmacHex,
      'kv': '$keyVersion',
    });
    final r = await _http.get(uri, headers: _headers);
    return VerifyResult.fromJson(_parse(r));
  }
}

class VerifyResult {
  VerifyResult({
    required this.chipAuthentic,
    required this.notReplayed,
    required this.issuerVerified,
    this.issuerId,
    this.reason = '',
  });

  final bool chipAuthentic;
  final bool notReplayed;
  final bool issuerVerified;
  final String? issuerId;
  final String reason;

  bool get fullyValid => chipAuthentic && notReplayed && issuerVerified;

  factory VerifyResult.fromJson(Map<String, dynamic> j) => VerifyResult(
        chipAuthentic: j['chip_authentic'] as bool,
        notReplayed: j['not_replayed'] as bool,
        issuerVerified: j['issuer_verified'] as bool,
        issuerId: j['issuer_id'] as String?,
        reason: (j['reason'] ?? '') as String,
      );
}

File: ts/veritag-sdk/src/attestClient.ts
/** HTTP client for the AttestCore web API. Domain-agnostic. */

export interface VerifyResult {
  chip_authentic: boolean;
  not_replayed: boolean;
  issuer_verified: boolean;
  issuer_id?: string | null;
  reason: string;
}

export interface ProvisionResult {
  uid: string;
  chip_key_hex: string;
  key_version: number;
  binding_id: string;
}

/** Thrown on any non-2xx response, carrying the status and the API's detail. */
export class AttestApiError extends Error {
  constructor(
    readonly status: number,
    readonly detail: string,
  ) {
    super(`AttestCore ${status}: ${detail}`);
    this.name = "AttestApiError";
  }
}

export class AttestClient {
  constructor(
    private readonly baseUrl: string,
    private readonly bearerToken?: string,
  ) {}

  private headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      ...(this.bearerToken ? { Authorization: `Bearer ${this.bearerToken}` } : {}),
    };
  }

  /** Parse a response, throwing AttestApiError on non-2xx instead of returning
   *  an error body cast as a success type. */
  private async parse<T>(r: Response): Promise<T> {
    const body = await r.json().catch(() => ({}));
    if (!r.ok) {
      const detail = (body as { detail?: string }).detail ?? r.statusText;
      throw new AttestApiError(r.status, detail);
    }
    return body as T;
  }

  /** Register an issuer. Returns its id AND the bearer token (shown once) needed
   *  to authenticate provisioning. */
  async registerIssuer(
    publicKeyHex: string,
    policy: Record<string, unknown> = {},
  ): Promise<{ issuerId: string; token: string }> {
    const r = await fetch(`${this.baseUrl}/v1/issuers`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ public_key_hex: publicKeyHex, policy }),
    });
    const b = await this.parse<{ issuer_id: string; token: string }>(r);
    return { issuerId: b.issuer_id, token: b.token };
  }

  /** Revoke an issuer (an issuer may only revoke itself). */
  async revokeIssuer(issuerId: string): Promise<void> {
    const r = await fetch(`${this.baseUrl}/v1/issuers/${encodeURIComponent(issuerId)}/revoke`, {
      method: "POST",
      headers: this.headers(),
    });
    await this.parse<{ status: string }>(r);
  }

  async provision(args: {
    uid: string;
    issuerId: string;
    bindingPayloadHex: string;
    signatureHex: string;
    keyVersion?: number;
  }): Promise<ProvisionResult> {
    const r = await fetch(`${this.baseUrl}/v1/tags/provision`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        uid: args.uid,
        issuer_id: args.issuerId,
        binding_payload_hex: args.bindingPayloadHex,
        signature_hex: args.signatureHex,
        key_version: args.keyVersion ?? 1,
      }),
    });
    return this.parse<ProvisionResult>(r);
  }

  async verify(uid: string, ctr: number, cmacHex: string, keyVersion = 1): Promise<VerifyResult> {
    const qs = new URLSearchParams({ u: uid, c: String(ctr), m: cmacHex, kv: String(keyVersion) });
    const r = await fetch(`${this.baseUrl}/v1/verify?${qs}`, { headers: this.headers() });
    return this.parse<VerifyResult>(r);
  }
}

export const fullyValid = (v: VerifyResult): boolean =>
  v.chip_authentic && v.not_replayed && v.issuer_verified;
